test: changing signature of wait_until().

rajarshimaitra opened the pull request:
As a follow up on [#18986(comment)](https://github.com/bitcoin/bitcoin/pull/18986#discussion_r426768706), jnewbery suggested that wait_until() is unnecessarily taking `timeout_factor` as a parameter, whereas the client should have this knowledge and can directly provide `wait_until()` the required timeout.

As it turned out almost nowhere the client is passing `timeout_factor` to `wait_until()` and changing the signature is enough to enforce the intended behavior.

Verified all the existing tests are passing. 
Verified all the existing tests are passing with `timeout-factor 0`.

fanquake commented:
You have committed your vscode settings as part of af83312ac57e6dfe2c7202db81ef54b7efd31170.

jonatack commented on test/functional/test_framework/mininode.py:
```diff
@@ -372,7 +372,7 @@ def on_version(self, message):
 
     def wait_until(self, test_function, timeout):
-        wait_until(test_function, timeout=timeout, lock=mininode_lock, timeout_factor=self.timeout_factor)
+        wait_until(test_function, timeout=timeout, lock=mininode_lock)
```
This belongs in the other commit?

Generally a typo isn't worth a separate commit. As you spend more time reviewing PRs, you'll see these practices.

rajarshimaitra replied:
Right, it's all mixed up. Fixed now.

jnewbery replied:
This breaks the factor argument. You need to call `wait_until()` with `timeout = timeout * self.timeout_factor`.

rajarshimaitra replied:
Agreed, fixed.

jonatack reviewed (commented):
Could you update the PR description to link directly to https://github.com/bitcoin/bitcoin/pull/18986#discussion_r426768706, the comment where this change was suggested, rather than to the PR? This avoids asking your reviewers to search through the whole PR discussion to find the suggestion.

Also, take your time to review your changes before opening a PR -- [don't rush](https://jonatack.github.io/articles/how-to-contribute-pull-requests-to-bitcoin-core#do-it-well) ;)

jonatack commented:
Also, you can see that a maintainer changed your PR description. Do not put GitHub @-prefixed usernames in commits and PR descriptions; the latter because usernames in the description are copied into the merge commit by the merge script. This can cause annoying notifications for those concerned. An update to the merge script now warns maintainers if a merge message contains a @username, but it's better not to add them to begin with. (This paragraph is from https://jonatack.github.io/articles/how-to-contribute-pull-requests-to-bitcoin-core)

rajarshimaitra committed:
test: change signature of wait_until()

wait_until() does not take timeout_factor as an arguement anymore.
Clients should pass in the correct timeout value at the time of calling.

rajarshimaitra commented:
Sorry for all those silly mistakes. Fixed them and rebased.

rajarshimaitra commented:
There is a Travis failure happening due to `feature_backwards_compatibility.py`. This test is skipping with the following warning (probably why I missed it). 
```
$ test/functional/feature_backwards_compatibility.py
2020-05-22T14:15:55.381000Z TestFramework (WARNING): Test Skipped: previous releases not available or disabled
2020-05-22T14:15:55.431000Z TestFramework (INFO): Stopping nodes
2020-05-22T14:15:55.432000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_8xy_o2kz
on exit
2020-05-22T14:15:55.432000Z TestFramework (INFO): Test skipped
```
Any pointer on how to enable the previous release?

jnewbery commented on test/functional/test_framework/test_node.py:
```diff
@@ -358,7 +358,7 @@ def is_node_stopped(self):
 
     def wait_until_stopped(self, timeout=BITCOIND_PROC_WAIT_TIMEOUT):
-        wait_until(self.is_node_stopped, timeout=timeout, timeout_factor=self.timeout_factor)
+        wait_until(self.is_node_stopped, timeout=timeout)
```
Do you need to multiply `timeout` by `timeout_factor` here?

rajarshimaitra replied:
Timeout here is set to `BITCOIND_PROC_WAIT_TIMEOUT` which is hardcode as 60 secs. This is the only time this value is being called. As it has been specifically separated from `self.options.timeout`, would it make sense to multiply it with `timeout_factor` here? Thinking that I skipped it.

rajarshimaitra replied:
Polite reminder. :) 
Does the above rationale make sense or should I change it?

jnewbery replied:
> Timeout here is set to BITCOIND_PROC_WAIT_TIMEOUT which is hardcode as 60 secs.

That's the default value. `wait_until_stopped()` can be called with any timeout.

rajarshimaitra replied:
Right that was just a default. agreed and fixed.

jnewbery commented on test/functional/test_framework/mininode.py:
```diff
@@ -205,7 +205,7 @@ def _on_data(self):
                 self.on_message(t)
         except Exception as e:
-            logger.exception('Error reading message:', repr(e))
+            logger.exception('Error reading message: %s' % (repr(e)))
```
Actually, why is this even being changed in this PR? It seems completely unrelated to the other changes.

rajarshimaitra replied:
Yes, this line was giving error. So I changed it as a nit.

laanwj replied:
The previous formulation was wrong, so it makes sense to change this. I also prefer `{}`.format() syntax though, and it's what the code base has been moving towards for a while.

rajarshimaitra replied:
Fixed.

rajarshimaitra commented:
Thank you all for the reviews. Updated as per the suggestions, rebased, and retested. This is ready for a second look.

gzhao408 commented on test/functional/test_framework/util.py:
```diff
@@ -208,10 +208,9 @@ def str_to_b64str(string):
     return Decimal(amount).quantize(Decimal('0.00000001'), rounding=ROUND_DOWN)
 
-def wait_until(predicate, *, attempts=float('inf'), timeout=float('inf'), lock=None, timeout_factor=1.0):
+def wait_until(predicate, *, attempts=float('inf'), timeout=float('inf'), lock=None):
```
Mininode has its own `wait_until` function which automatically grabs the `mininode_lock` (in addition to including the mininode's `timeout_factor`) so I think it’s probably better to use that function if synchronization is needed.
Would it be worth adding a comment about this? I’ve seen multiple reviews with reminders to pass in the `mininode_lock` or use another function that does it automatically. I wonder if a comment would help remind people earlier on.

rajarshimaitra replied:
Thank @gzhao408 for the review and testing. While I agree with the cautionary message of using the mininode `wait_until`, I am not sure whether this is the correct place to add that comment. This seems like a generalized function for `wait_until` that can be used in many places and not all of them might require the lock. 
It seems that comment can be placed in the mininode's definition itself as a reminder to use `mininode.wait_until` whenever required as opposed to `util.wait_until`.

gzhao408 replied:
Well, no, my suggestion for the comment is more of a "don't use this." 
I don't think `util.wait_until` should ever be used except to define object-specific `wait_until` since they should be responsible for including the `timeout_factor` (that's the point of this PR right?). In functional tests, I think it's more appropriate to use mininode, test_node, or test_framework `wait_until`. See #19134 - your PRs are very intertwined.

rajarshimaitra replied:
Hmm I see your point now, a comment here saying "Dont Use it" can help as per #19134 modifications. I am also now confused about why there are so many wait_untils to begin with. Anyway, I am not very sure if this PR is adding any value. `timeout_factor` is not being used anywhere with wait_until anyway. I think I will wait for more ACKs before further working on this.

glozow reviewed (commented):
Looks good to me, tests passed and I did a code review for the places using `test_framework.util`'s `wait_until` function. I left a comment because I think other `wait_until` functions are more favorable due to synchronization needs as well as `timeout_factor`. 
I also have a conceptual question that hopefully doesn't sound offensive/dumb: I don't really see many examples of `timeout_factor` being used - what are the use cases for it?

rajarshimaitra commented:
> I also have a conceptual question that hopefully doesn't sound offensive/dumb: I don't really see many examples of `timeout_factor` being used - what are the use cases for it?

I also have the same question. Almost all of the places `timeout` has been explicitly passed into `wait_until` and it was using the default factor of 1. Thus the change became so small and didn't require modification in those caller functions.

One reason I can think of is to easily bump up all the timeouts at once. PR [#18716](https://github.com/bitcoin/bitcoin/pull/18617) that added this flag which was motivated by this [comment](https://github.com/bitcoin/bitcoin/pull/18249#issuecomment-593944985).

It seems it was added to solve an issue rather than a functionality. Maybe someone else can shed more light. Curious as well.

fanquake commented:
@MarcoFalke can you concept ack / nack here? @jnewbery can you comment re https://github.com/bitcoin/bitcoin/pull/19048#issuecomment-636473840?

jnewbery commented:
> @jnewbery can you comment

I think this can just be closed. It's required rebase for several months.
